src/lightning/data/streaming/shuffle.py

```diff
@@ -0,0 +1,167 @@
+        self.random_state = np.random.RandomState(seed=self.seed + current_epoch)  # type: ignore
+        chunk_intervals = self.cache.get_chunk_intervals()
+        indexes = list(range(len(chunk_intervals)))
+        shuffled_chunk_intervals = np.asarray(chunk_intervals)[indexes]
+
+        chunks_per_process: List[List[int]] = [[] for _ in range(distributed_env.world_size)]
+        intervals_per_process: List[List[List[int]]] = [[] for _ in range(distributed_env.world_size)]
+        for index, (chunk_index, chunk_interval) in enumerate(zip(indexes, shuffled_chunk_intervals)):
+            replica_index = index % distributed_env.world_size
+            chunks_per_process[replica_index].append(chunk_index)
+            intervals_per_process[replica_index].append(chunk_interval)
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        return array.tolist()
+
+
+class MinShuffle(Shuffle):
+    """MinShuffle shuffle the items and ensure all the processes receive the same number of items."""
+
+    @lru_cache(maxsize=10)
+    def get_len(self, distributed_env: _DistributedEnv, current_epoch: int) -> int:
+        _, intervals_per_process = self.get_chunks_and_intervals_per_process(distributed_env, current_epoch)
+        min_items_per_process = min(
+            [sum([(interval[-1] - interval[0]) for interval in intervals]) for intervals in intervals_per_process]
+        )
+        return min_items_per_process
+
+    @lru_cache(maxsize=10)
+    def get_chunks_and_intervals_per_process(self, distributed_env: _DistributedEnv, current_epoch: int) -> Any:
+        self.random_state = np.random.RandomState(seed=self.seed + current_epoch)  # type: ignore
+        chunk_intervals = self.cache.get_chunk_intervals()
+        indexes = range(len(chunk_intervals))
+        shuffled_indexes = self.random_state.permutation(indexes)
+        shuffled_chunk_intervals = np.asarray(chunk_intervals)[shuffled_indexes]
+
+        chunks_per_process: List[List[int]] = [[] for _ in range(distributed_env.world_size)]
+        intervals_per_process: List[List[List[int]]] = [[] for _ in range(distributed_env.world_size)]
+        for index, (chunk_index, chunk_interval) in enumerate(zip(shuffled_indexes, shuffled_chunk_intervals)):
+            replica_index = index % distributed_env.world_size
+            chunks_per_process[replica_index].append(chunk_index)
+            intervals_per_process[replica_index].append(chunk_interval)
+
+        return chunks_per_process, intervals_per_process
+
+    def __call__(self, array: np.ndarray) -> List[int]:
+        assert self.random_state
+        return self.random_state.permutation(array).tolist()
+
+
+class AcrossChunkShuffle(Shuffle):
+    """AcrossChunkShuffle shuffle the items and ensure all the processes receive the same number of items."""
```

Comment:
This is a repetition of the previous docstring.

Reply:
done.